File: src/main/java/org/yearobjectives/infrastructure/database/repository/AccomplishmentRepository.java
package org.yearobjectives.infrastructure.database.repository;

import io.quarkus.hibernate.orm.panache.PanacheRepository;
import jakarta.enterprise.context.ApplicationScoped;
import org.yearobjectives.infrastructure.database.entity.AccomplishmentEntity;

import java.util.List;
import java.util.UUID;

@ApplicationScoped
public class AccomplishmentRepository implements PanacheRepository<AccomplishmentEntity> {

    public List<AccomplishmentEntity> getCurrentAccomplishments(final UUID objectiveId, final Long currentTimeStamp) {
        return list("objective.resourceId = ?1 and maxPossibleStart > ?2 and minPossibleStart <= ?2", objectiveId, currentTimeStamp);
    }
}

File: src/main/kotlin/com/objectives/management/infrastructure/database/repository/AccomplishmentRepository.kt
package com.objectives.management.infrastructure.database.repository

import com.objectives.management.infrastructure.database.model.AccomplishmentEntity
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Repository
import java.util.*

@Repository
interface AccomplishmentRepository: JpaRepository<AccomplishmentEntity, Long> {

    fun findByObjectiveResourceIdAndMaxPossibleStartGreaterThanAndMinPossibleStartLessThanEqual(
        resourceId: UUID,
        maxTimestamp: Long,
        minTimestamp: Long
    ): List<AccomplishmentEntity>

    fun findByResourceId(accomplishmentId: UUID): AccomplishmentEntity?
}

fun AccomplishmentRepository.findCurrentByObjective(
    objectiveId: UUID,
    currentTimestamp: Long
): List<AccomplishmentEntity> {
    return findByObjectiveResourceIdAndMaxPossibleStartGreaterThanAndMinPossibleStartLessThanEqual(
        objectiveId,
        currentTimestamp,
        currentTimestamp
    )
}